### PR TITLE
[mt8186-v02] waves: store config blob in a cache in waves.c

### PR DESCRIPTION
**Pull request by barry-waves**
Store/apply config blob in a cache to avoid that
cfg.data will be released after prepare.

### CODE REVIEW

File: src/audio/module_adapter/module/waves.c [hunk 1]
```diff
@@ -596,95 +652,77 @@ static int waves_codec_init(struct processing_module *mod)
...
-		return -ENOMEM;
+		ret = -ENOMEM;
+		goto err;
 	}
 	waves_codec->response = response;
 
 	comp_dbg(dev, "waves_codec_init() done");
 	return ret;
+
+err:
```
Comment (btian1) [inline src/audio/module_adapter/module/waves.c:691]: please double check, you may missed free more memory here, there should have multiple err label here:

err1:

err2:

err3:

for example: waves_codec->effect memory?

File: src/audio/module_adapter/module/waves.c [hunk 2]
```diff
@@ -596,95 +652,77 @@ static int waves_codec_init(struct processing_module *mod)
...
+	return ret;
 }
 
 static int waves_codec_prepare(struct processing_module *mod)
 {
 	struct comp_dev *dev = mod->dev;
 	int ret;
 
-	comp_dbg(dev, "waves_codec_prepare() start");
+	comp_info(dev, "waves_codec_prepare() start");
```
Comment (btian1) [inline src/audio/module_adapter/module/waves.c:703]: comp_dbg

File: src/audio/module_adapter/module/waves.c [hunk 3]
```diff
@@ -462,6 +464,51 @@ static int waves_effect_revision(struct processing_module *mod)
 	return 0;
 }
 
+/* cache config blob*/
+static int waves_effect_save_config_blob_to_cache(struct processing_module *mod,
```
[the remark this answers is not in the document]
Comment (barry-waves) [reply src/audio/module_adapter/module/waves.c:468]: The cache is allocated by using module_allocate_memory()

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/module_adapter/module/waves.c:468]: ok, so its a local memory cache and not a DSP HW data cache.

File: src/audio/module_adapter/module/waves.c [hunk 4]
```diff
@@ -603,64 +659,44 @@ static int waves_codec_init(struct processing_module *mod)
...
-		return -EINVAL;
+		comp_err(dev, "cadence_codec_init() MaxxEffect_GetMessageMaxSize returned %d", ret);
+		ret = -EINVAL;
+		goto err1;
 	}
 
 	response = module_allocate_memory(mod, waves_codec->response_max_bytes, 16);
 	if (!response) {
-		comp_err(dev, "waves_codec_init() failed to allocate %d bytes for response",
+		comp_err(dev, "cadence_codec_init() failed to allocate %d bytes for response",
```
Comment (lgirdwood) [inline src/audio/module_adapter/module/waves.c:682]: cadence ?

**Review by lgirdwood — COMMENTED** [review]
btw, feel free to follow up with another PR if you like that changes the API naming to effect instead of codec.

**Comment by barry-waves** [comment]
Hi @lgirdwood , @btian1 and @Zames-Chang ,
Could you please review it? Thank you.

**Comment by lgirdwood** [comment]
> Hi @lgirdwood , @btian1 and @Zames-Chang , Could you please review it? Thank you.

@barry-waves this is MTK production branch so need @thesofproject/mediatek  to review and merge. 

@andyross @cujomalainey fyi.

File: src/audio/module_adapter/module/waves.c [hunk 5]
```diff
@@ -596,95 +652,77 @@ static int waves_codec_init(struct processing_module *mod)
 	int ret = 0;
 	void *response = NULL;
 
-	comp_dbg(dev, "waves_codec_init() start");
+	comp_info(dev, "waves_codec_init() start");
```
[the remark this answers is not in the document]
Comment (cujomalainey) [reply src/audio/module_adapter/module/waves.c]: @btian1 i would agree, but we should also avoid modifying cherry picks any more than necessary

**Review by btian1 — APPROVED** [review]
if you'd like, you can follow up with another PR, move waves out of module_adapter folder.
and move it to under codec folder.

**Comment by barry-waves** [comment]
Hi @Zames-Chang and @thesofproject/mediatek ,
Could you please merge this patch? Thank you.

**Review by chunxu-li — APPROVED** [review]

**Comment by lgirdwood** [comment]
@andyross @cujomalainey fyi - somebody from MTK needs to merge.

**Comment by cujomalainey** [comment]
> @andyross @cujomalainey fyi - somebody from MTK needs to merge.

Why does MTK need to merge? The code change is entirely within the Waves block.

**Comment by barry-waves** [comment]
Hi,
This pull request cherry-picks from main to mt8186/v0.2

**Comment by johnylin76** [comment]
This commit needs to be landed on mt8186/v0.2 as a bug fix for the existing Waves-integrated CrOS projects (later than TigerLake precisely, i.e. MT8186, ADL-P/N, RPL.

**Comment by barry-waves** [comment]
Hi @Zames-Chang and @thesofproject/mediatek ,
Could you please merge this patch? Thank you.

**Comment by barry-waves** [comment]
Hi @Zames-Chang and @thesofproject/mediatek ,
Could you please merge this pull request? Thank you.

**Pull request merged**